services/grpc/lightning.methods.js

```diff
@@ -260,73 +260,64 @@ async function closeChannel(payload = {}) {
     try {
       const call = this.service.sendPayment(payload)
 
       call.on('data', data => {
-        const isSuccess = !data.payment_error
-        if (isSuccess) {
-          grpcLog.debug('PAYMENT SUCCESS', data)
-
-          // Convert payment_hash to hex string.
-          let paymentHash = data.payment_hash
-          if (paymentHash) {
-            paymentHash = paymentHash.toString('hex')
-          }
-
-          // In some cases lnd does not return the payment_hash. If this happens, retrieve it from the invoice.
-          else {
-            const invoice = bolt11DecodePayReq(payload.payment_request)
-            const paymentHashTag = invoice.tags
-              ? invoice.tags.find(t => t.tagName === 'payment_hash')
-              : null
-            paymentHash = paymentHashTag ? paymentHashTag.data : null
-          }
-
-          // Convert the preimage to a hex string.
-          const paymentPreimage = data.payment_preimage
-            ? data.payment_preimage.toString('hex')
+        // Convert payment_hash to hex string.
+        data.payment_hash = data.payment_hash ? data.payment_hash.toString('hex') : null
+
+        // In some cases lnd does not return the payment_hash. If this happens, retrieve it from the invoice.
+        if (!data.payment_hash) {
+          const invoice = bolt11DecodePayReq(payload.payment_request)
+          const paymentHashTag = invoice.tags
+            ? invoice.tags.find(t => t.tagName === 'payment_hash')
             : null
+          data.payment_hash = paymentHashTag ? paymentHashTag.data : null
+        }
+
+        // Convert the preimage to a hex string.
+        data.payment_preimage = data.payment_preimage ? data.payment_preimage.toString('hex') : null
+
+        // Add lnd return data, as well as payment preimage and hash as hex strings to the response.
+        Object.assign(res, data)
 
-          // Notify the client of a successful payment.
-          const res = {
-            ...payload,
-            data,
-            payment_preimage: paymentPreimage,
-            payment_hash: paymentHash,
-          }
-          this.emit('sendPayment.data', res)
+        // Payment success is determined by the absense of a payment error.
+        const isSuccess = !res.payment_error
+        if (isSuccess) {
+          grpcLog.debug('PAYMENT SUCCESS', res)
           resolve(res)
-        } else {
-          // Notify the client if there was a problem sending the payment
-          grpcLog.error('PAYMENT ERROR', data)
-          const error = new Error(data.payment_error)
-          error.payload = payload
-          this.emit('sendPayment.error', error)
-          reject(error)
         }
 
+        // In case of an error, notify the client if there was a problem sending the payment.
+        else {
+          grpcLog.error('PAYMENT ERROR', res)
+          const error = new Error(res.payment_error)
+          error.details = res
+          reject(error)
```

Comment:
Is there a reason why here we reject and in `catch` we re-throw?

Reply:
The rejection here is used to reject the Promise within the `data` event handler in the case that the `sendPayment` lightning call returns an error (which it does by returning a success but with `payment_error` set).

The catch at the end wraps the entire method and is triggered in the event that an unexpected error occurs anywhere within our `sendPayment` wrapper. It doesn't re throw the rejection from the `data` event handler.